gems/sorbet-runtime/test/types/props/constructor.rb

```diff
@@ -145,4 +145,76 @@ def initialize(hash={})
     end
     assert_equal("Missing required prop `name` for class `Opus::Types::Test::Props::ConstructorTest::CustomInitializeStruct`", err.message)
   end
+
+  class ComplexStruct < T::Struct
+    prop :primitive, Integer
+    prop :nilable, T.nilable(Integer)
+    prop :nilable_on_read, T.nilable(Integer), raise_on_nil_write: true
+    prop :primitive_default, Integer, default: 0
+    prop :primitive_nilable_default, T.nilable(Integer), default: 0
+    prop :factory, Integer, factory: -> {0}
+    prop :primitive_array, T::Array[Integer]
+    prop :array_default, T::Array[Integer], default: []
+    prop :primitive_hash, T::Hash[String, Integer]
+    prop :array_of_nilable, T::Array[T.nilable(Integer)]
+    prop :nilable_array, T.nilable(T::Array[Integer])
+    prop :substruct, UntypedField
+    prop :nilable_substract, T.nilable(UntypedField)
+    prop :default_substruct, UntypedField, default: UntypedField.new
+    prop :array_of_substruct, T::Array[UntypedField]
+    prop :hash_of_substruct, T::Hash[String, UntypedField]
+    prop :custom_type, Inner
+    prop :nilable_custom_type, T.nilable(Inner)
+    prop :default_custom_type, Inner, default: Inner.new(i: 4)
+    prop :array_of_custom_type, T::Array[Inner]
+    prop :hash_of_custom_type_to_substruct, T::Hash[Inner, UntypedField]
+    prop :unidentified_type, Object
+    prop :nilable_unidentified_type, T.nilable(Object)
+    prop :array_of_unidentified_type, T::Array[Object]
+    prop :defaulted_unidentified_type, Object, default: Object.new
+    prop :hash_with_unidentified_types, T::Hash[Object, Object]
+    prop :infinity_float, Float, default: Float::INFINITY
+    prop :setter_validate, T.nilable(Object), setter_validate: -> {}
+    prop :raise_on_nil_write, T.nilable(Object), raise_on_nil_write: true
+  end
+
+  describe 'generated code' do
+    describe 'initialize' do
+      it 'validates' do
+        src = ComplexStruct.decorator.send(:generate_initialize_source).to_s
+        T::Props::GeneratedCodeValidation.validate_initialize(src)
+      end
+
+      it 'has meaningful validation which complains at lurky method invocation' do
+        src = ComplexStruct.decorator.send(:generate_initialize_source).to_s
+        src = src.sub(/\.deep_clone_object\b/, '.something_suspicious')
+        assert_raises(T::Props::GeneratedCodeValidation::ValidationError) do
+          T::Props::GeneratedCodeValidation.validate_initialize(src)
+        end
+      end
+    end
+
+    describe 'disabling evaluation' do
```

Comment:
I think there also needs to be a test that:

* defines props, then defines `initialize`
* defines `initialize`, then defines props
* defines props, then defines `initialize`, then defines more props

I think the correct behavior would be to call the user's `initialize` in all cases.  (I think the right thing already happens?  But I'm not entirely sure.)  This would be another good candidate for a separate PR.